cis194/week4/zehua/src/Exercises.hs

```diff
@@ -25,3 +26,64 @@ fun2' = foldr (+) 0
     where
       oddv = (+1) . (*3)
       evenv = (`div` 2)
+
+
+-- ex2
+data Tree a = Leaf
+            | Node Integer (Tree a) a (Tree a)
+  deriving (Show, Eq)
+
+getHeight :: Tree a -> Integer
+getHeight Leaf            = -1
+getHeight (Node h1 _ _ _) = h1
+
+foldTree :: [a] -> Tree a
+foldTree = foldr f Leaf
+  where
+    f x Leaf = Node 0 Leaf x Leaf
+    f x (Node h l n r)
+        | hl == hr  = let l'  = f x l
+                          hl' = getHeight l'
+                          h'  = bool h (h+1) (hl < hl')
+                      in Node h' l' n r
+        | hl < hr   = Node h (f x l) n r
+        | otherwise = Node h l n (f x r)
+      where
+        hl = getHeight l
+        hr = getHeight r
+
+
+-- ex3
+xor2 :: Bool -> Bool -> Bool
+xor2 = (/=)
+
+xor :: [Bool] -> Bool
+xor = foldr xor2 False
+
+
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr ((:) . f) []
+
+
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f base = foldr (flip f) base . reverse
```

Comment:
There are 2 folding happen here, one is `reverse` one is `foldr`. Can we use only 1 folding?
hint: What about function as accumulator for foldr? because nested function calls when evaluated is in the FILO order.

Reply:
thanks for the suggestion. updated the code